pilot/pkg/networking/plugin/authz/rbac_v2.go

```diff
@@ -0,0 +1,172 @@
+		if authzPolicy.Policy.WorkloadSelector != nil && !service.isServiceMatchedWorkload(authzPolicy.Policy.WorkloadSelector.Labels) {
+			continue
+		}
+		for i, subject := range authzPolicyToSubjects[authzPolicy.Name] {
+			// Convert role.
+			role := option.authzPolicies.GetServiceRoleFromName(namespace, subject.RoleRef.Name)
+			permission := &policyproto.Permission{}
+			rule := role.Rules[0]
+
+			// Check to make sure that the current service (caller) is the one this rule is applying to.
+			if !service.match(rule) {
+				continue
+			}
+			// Also check if we have converted this rule before.
+			if !hasConvertedPermission(rbac, subject.RoleRef.Name) {
+				rbacLog.Debugf("rules[%d] matched", i)
+				if option.forTCPFilter {
+					// TODO(yangminzhu): Move the validate logic to push context and add metrics.
+					if err := validateRuleForTCPFilter(rule); err != nil {
+						// It's a user misconfiguration if a HTTP rule is specified to a TCP service.
+						// For safety consideration, we ignore the whole rule which means no access is opened to
+						// the TCP service in this case.
+						rbacLog.Debugf("rules[%d] ignored, found HTTP only rule for a TCP service: %v", i, err)
+						continue
+					}
+				}
+				// Generate the policy if the service is matched and validated to the services specified in
+				// ServiceRole.
+				permission = convertToPermission(rule)
+				if permission.String() == "" {
+					rbacLog.Debugf("role %s skipped for no rule matched", subject.RoleRef.Name)
+					continue
+				}
+			}
+
+			if option.forTCPFilter {
+				if err := validateBindingsForTCPFilter([]*rbacproto.ServiceRoleBinding{subject}); err != nil {
+					rbacLog.Debugf("role %s skipped, found HTTP only binding for a TCP service: %v", subject.RoleRef.Name, err)
+					continue
+				}
+			}
+			// Convert binding.
+			enforcedPrincipals, _ := convertToPrincipals([]*rbacproto.ServiceRoleBinding{subject}, option.forTCPFilter)
+
+			if len(enforcedPrincipals) == 0 {
+				rbacLog.Debugf("role %s skipped for no principals found", subject.RoleRef.Name)
+				continue
+			}
+
+			if len(enforcedPrincipals) != 0 {
+				if rbac.Policies[subject.RoleRef.Name] == nil {
+					rbac.Policies[subject.RoleRef.Name] = &policyproto.Policy{
+						Permissions: []*policyproto.Permission{permission},
+						Principals:  enforcedPrincipals,
+					}
+				} else {
+					allPrincipals := rbac.Policies[subject.RoleRef.Name].Principals
+					// Check if enforcedPrincipals has existed in the allPrincipals.
+					if !isPrincipalInPrincipals(enforcedPrincipals, allPrincipals) {
+						rbac.Policies[subject.RoleRef.Name].Principals = append(allPrincipals, enforcedPrincipals...)
+					}
+				}
+			}
+		}
+	}
+	return &http_config.RBAC{Rules: rbac}
+}
+
+// isServiceMatchedWorkload checks if the calling service is matched with the selector in the AuthorizationPolicy if
+// present.
+func (service serviceMetadata) isServiceMatchedWorkload(labels map[string]string) bool {
```

Comment:
// IsSupersetOf checks if the service's labels are super set of the given label.
`func (service serviceMetadata) IsSupersetOf(label map[string]string) bool {`